README.md
# Google Auth Token Validator

[![Build Status](https://travis-ci.org/hamza/google-auth-token-validator-ruby.svg?branch=master)](https://travis-ci.org/hamza/google-auth-token-validator-ruby)
[![Coverage Status](https://coveralls.io/repos/github/hamza/google-auth-token-validator-ruby/badge.svg?branch=master)](https://coveralls.io/github/hamza/google-auth-token-validator-ruby?branch=master)

The Google Sign-In API gives OAuth2 JSON Web Tokens (JWT) as response data upon user sign-in. A necessary step for a service provider to trust such a token is validatation. Accepting the token without validation would allow a malicious client to simply assert itself in your system.

Google provides [libraries in several languages](https://developers.google.com/identity/sign-in/web/backend-auth) to accomplish this, as well as an API endpoint that can outsource the task to Google's own servers (thereby introducing an additional network round trip into every authentication step), but a Ruby implementation is missing. This gem fills that gap.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'google-auth-token_validator'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install google-auth-token_validator

## Usage

```ruby
require "googleauth/token_validator"

client_id = "<your client ID>"
id_token = "<user's ID token>"

begin
  valid = Google::Auth::TokenValidator.new(id_token, client_id).validate
rescue Google::Auth::TokenValidator::Error => e
  puts e.message
  valid = false
end

if valid
  # trust this token
end
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

This gem uses [`dotenv`](https://github.com/bkeepers/dotenv) to set up the test environment. To enable local testing, you'll need to set up the following -

|Key|Description|
|---|---|
|`TEST_CLIENT_ID`|A Google OAuth client ID, used to compare ID tokens against|
|`TEST_CLIENT_SECRET`|A Google OAuth client secret, used to authorize token refreshes|
|`REFRESH_TOKEN`|A refresh token for a test user. It is recommended to use the [Google OAuth Playground](https://developers.google.com/oauthplayground/) to generate one|

**NOTE**: it is not recommended to use your real account for this. Instead use a test user.

Use the [Google Cloud Console](https://console.cloud.google.com) to set up an OAuth id/secret pair for use with testing.

## Contributing

Bug reports and pull requests are welcome on [the repo](https://github.com/hamza/google-auth-token-validator-ruby/issues).

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
# 0.1.2 - September 17, 2017

- make description more clear

# 0.1.1 - September 17, 2017

- change module nesting syntax (fix)

# 0.1.0 - September 17, 2017

- initial implementation


spec/spec_helper.rb
require "coveralls"
Coveralls.wear!

require "bundler/setup"
require "yaml"
require "json"
require "googleauth/token_validator"

unless ENV["TRAVIS"]
  require "dotenv"
  Dotenv.load File.join(__dir__, "..", ".env")
end

RSpec.shared_context "test tokens", :shared_context => :metadata do
  yaml = YAML.load_file File.join(__dir__, "test_tokens.yml")
  let(:tokens) { JSON.parse(yaml.to_json, symbolize_names: true) }
end

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end

  config.include_context "test tokens", :include_shared => true
end


spec/google-auth_token_validator_spec.rb
require "spec_helper"
require "net/http"
require "uri"
require "pry"

TOKEN_REFRESH_URL = URI("https://www.googleapis.com/oauth2/v4/token")

RSpec.describe Google::Auth::TokenValidator do
  include_context "test tokens"

  before :all do
    @client_id = ENV["TEST_CLIENT_ID"]

    response = Net::HTTP.post_form(
      TOKEN_REFRESH_URL,
      {
        client_id: @client_id,
        client_secret: ENV["TEST_CLIENT_SECRET"],
        refresh_token: ENV["REFRESH_TOKEN"],
        grant_type: "refresh_token"
      }
    )

    @refreshed_id_token = JSON.parse(response.body, symbolize_names: true)[:id_token]
  end

  before :each do
    @validator = Google::Auth::TokenValidator.new @refreshed_id_token, @client_id
  end

  it "has a version number" do
    expect(Google::Auth::TokenValidator::VERSION).not_to be nil
  end

  it "validates a token" do
    expect(@validator.validate).to be(true)
  end

  it "validates a token when passed multiple client IDs" do
    client_ids = [@client_id, "additional-client-ID"]
    validator = Google::Auth::TokenValidator.new @refreshed_id_token, client_ids

    expect(validator.validate).to be(true)
  end

  it "fails when initialized w/out a token consisting of 3 period-delimited segments" do
    expect {
      Google::Auth::TokenValidator.new "not.token", @client_id
    }.to raise_error(Google::Auth::TokenValidator::Error, /Wrong number of segments in token/)
  end

  %i(@signed @signature @envelope @payload).each do |param|
    it "fails when #{param} param is missing" do
      @validator.instance_variable_set(param, nil)

      expect {
        @validator.validate
      }.to raise_error(Google::Auth::TokenValidator::Error, "Validator was not properly initialized")
    end
  end

  it "fails when envelope contains non-matching certificate" do
    validator = Google::Auth::TokenValidator.new tokens[:no_matching_cert], @client_id

    expect {
      validator.validate
    }.to raise_error(
      Google::Auth::TokenValidator::Error,
      /No matching Google cert found for envelope/
    )
  end

  it "fails when the template is used too late" do
    validator = Google::Auth::TokenValidator.new tokens[:expired], @client_id

    expect {
      validator.validate
    }.to raise_error(Google::Auth::TokenValidator::Error, "Token used too late")
  end

  it "fails when the client ID is of invalid type" do
    validator = Google::Auth::TokenValidator.new @refreshed_id_token, :bogus_client_id

    expect {
      validator.validate
    }.to raise_error(Google::Auth::TokenValidator::Error, "Invalid client id(s)")
  end

  it "fails when the expected client ID doesn't match the provided one" do
    validator = Google::Auth::TokenValidator.new @refreshed_id_token, "not-a-real-client-ID"

    expect {
      validator.validate
    }.to raise_error(
      Google::Auth::TokenValidator::Error,
      "Wrong recipient - payload audience doesn't match required audience"
    )
  end

  it "fails when tampering has invalidated the signature" do
    validator = Google::Auth::TokenValidator.new tokens[:tampered], @client_id

    expect {
      validator.validate
    }.to raise_error(Google::Auth::TokenValidator::Error, "Token signature invalid")
  end

  # NOTE - this test should be run last, since it pollutes the validator class. This technique isn't
  # ideal, but I cannot actually trigger this check otherwise; tampering with the payload issuer
  # causes the signature check to fail prior to this.
  it "fails when issuer doesn't match expectations" do
    Google::Auth::TokenValidator::ISSUERS = %w(https://bogus.com).freeze
    validator = Google::Auth::TokenValidator.new @refreshed_id_token, @client_id

    expect {
      validator.validate
    }.to raise_error(Google::Auth::TokenValidator::Error, /Invalid issuer/)
  end
end


spec/google_signin_token_validator_spec.rb
require "spec_helper"

RSpec.describe Google::Auth::TokenValidator do
  it "has a version number" do
    expect(Google::Auth::TokenValidator::VERSION).not_to be nil
  end

  it "does something useful" do
    expect(false).to eq(true)
  end
end


lib/google_signin_token_validator.rb
require "net/http"
require "json"
require "base64"
require "openssl"

require "google_signin_token_validator/version"

module GoogleSigninTokenValidator
  GOOGLE_SIGNON_CERTS_URL    = "https://www.googleapis.com/oauth2/v1/certs".freeze
  ISSUERS                    = %w(accounts.google.com https://accounts.google.com).freeze
  CLOCK_SKEW_SECONDS         = 300 # five minutes
  MAX_TOKEN_LIFETIME_SECONDS = 86400 # one day

  attr_reader :signed, :signature, :envelope, :payload, :client_id

  def initialize jwt, client_id
    segments = jwt.split "."

    fail Error, "Wrong number of segments in token: #{jwt}" unless segments.size.eql? 3

    @client_id = client_id

    @signed = "#{segments[0]}.#{segments[1]}"
    @signature = segments[2]

    @envelope = JSON.parse Base64.decode64(segments[0])
    @payload = JSON.parse Base64.decode64(segments[1])
  end

  def validate max_expiry = MAX_TOKEN_LIFETIME_SECONDS
    unless [@signed, @signature, @envelope, @payload].all? &:present?
      fail Error, "Validator was not properly initialized"
    end

    unless @envelope["kid"].in? self.class._certs.keys
      fail Error, "No matching Google cert found for envelope: #{@envelope}"
    end

    pem = self.class._certs[envelope["kid"]]
    cert = OpenSSL::X509::Certificate.new pem
    digest = OpenSSL::Digest::SHA256.new

    verified = cert.public_key.verify(digest, Base64.urlsafe_decode64(@signature), @signed)

    fail Error, "Token signature invalid" unless verified

    fail Error, "No issue time in token" if @payload["iat"].blank?
    fail Error, "No expiration time in token" if @payload["exp"].blank?

    now      = Time.now.to_i
    earliest = @payload["iat"] - CLOCK_SKEW_SECONDS
    latest   = @payload["exp"] + CLOCK_SKEW_SECONDS

    fail Error, "Expiration time too far in future" if @payload["exp"] >= Time.now.to_i + max_expiry
    fail Error, "Token used too early" if now < earliest
    fail Error, "Token used too late" if now > latest

    unless @payload["iss"].in? ISSUERS
      fail Error, "Invalid issuer. Expected one of #{ISSUERS}, but got #{@payload["iss"]}"
    end

    fail Error, "Invalid client id(s)" unless @client_id.class.in? [Array, String]
    aud_verified = @payload["aud"].send({ Array => :in?, String => :eql? }[@client_id.class], @client_id)

    fail Error, "Wrong recipient - payload aud doesn't match required aud" unless aud_verified

    # if we made it this far without blowing up, token is valid
    return true
  end

  class << self
    attr_accessor :_cached_certs

    def _certs
      @_cached_certs ||= begin
        uri = URI(GOOGLE_SIGNON_CERTS_URL)
        JSON.parse Net::HTTP.get(uri)
      end
    end
  end

  class Error < StandardError; end
end


lib/googleauth/token_validator.rb
require "net/http"
require "json"
require "base64"
require "openssl"

module Google
  module Auth
    class TokenValidator
      GOOGLE_SIGNON_CERTS_URL    = "https://www.googleapis.com/oauth2/v1/certs".freeze
      ISSUERS                    = %w(accounts.google.com https://accounts.google.com).freeze
      CLOCK_SKEW_SECONDS         = 300 # five minutes
      MAX_TOKEN_LIFETIME_SECONDS = 86400 # one day

      attr_reader :signed, :signature, :envelope, :payload, :client_id

      def initialize jwt, client_id
        segments = jwt.split(".")

        fail Error, "Wrong number of segments in token: #{jwt}" unless segments.size.eql? 3

        @client_id = client_id

        @signed = "#{segments[0]}.#{segments[1]}"
        @signature = segments[2]

        @envelope = JSON.parse Base64.decode64(segments[0])
        @payload = JSON.parse Base64.decode64(segments[1])
      end

      def validate max_expiry = MAX_TOKEN_LIFETIME_SECONDS
        unless [@signed, @signature, @envelope, @payload].all?
          fail Error, "Validator was not properly initialized"
        end

        unless self.class._certs.keys.include? @envelope["kid"]
          fail Error, "No matching Google cert found for envelope: #{@envelope}"
        end

        pem = self.class._certs[envelope["kid"]]
        cert = OpenSSL::X509::Certificate.new pem
        digest = OpenSSL::Digest::SHA256.new

        verified = cert.public_key.verify(digest, Base64.urlsafe_decode64(@signature), @signed)

        fail Error, "Token signature invalid" unless verified

        fail Error, "No issue time in token" if @payload["iat"].to_s.empty?
        fail Error, "No expiration time in token" if @payload["exp"].to_s.empty?

        now      = Time.now.to_i
        earliest = @payload["iat"] - CLOCK_SKEW_SECONDS
        latest   = @payload["exp"] + CLOCK_SKEW_SECONDS

        fail Error, "Expiration time too far in future" if @payload["exp"] >= Time.now.to_i + max_expiry
        fail Error, "Token used too early" if now < earliest
        fail Error, "Token used too late" if now > latest

        unless ISSUERS.include? @payload["iss"]
          fail Error, "Invalid issuer. Expected one of #{ISSUERS}, but got #{@payload["iss"]}"
        end

        if @client_id.is_a? Array
          aud_verified = @client_id.include?(@payload["aud"])
        elsif @client_id.is_a? String
          aud_verified = @payload["aud"] == @client_id
        else
          fail Error, "Invalid client id(s)"
        end

        unless aud_verified
          fail Error, "Wrong recipient - payload audience doesn't match required audience"
        end

        return true
      end

      class << self
        attr_accessor :_cached_certs

        def _certs
          @_cached_certs ||= begin
            uri = URI(GOOGLE_SIGNON_CERTS_URL)
            JSON.parse Net::HTTP.get(uri)
          end
        end
      end

      class Error < StandardError; end
    end
  end
end
